### DISCUSSION TITLE
SimpleDrums still missing when linked dynamically

### DISCUSSION THREAD

**Discussion by kybos** [muse-sequencer/muse#1204] [Team Posts]
Hi Tim,

I've just noticed that simpledrums is still not there when MusE is compiled without MODULES_BUILD_STATIC.
The error is the same as before:

Checking file: </usr/local/lib/muse-4.0/synthi/simpledrums.so>

pluginScan: Standard error output from scan:
loadPluginLib: filename:/usr/local/lib/muse-4.0/synthi/simpledrums.so
muse_plugin_scan: dlopen(/usr/local/lib/muse-4.0/synthi/simpledrums.so) failed: /usr/local/lib/muse-4.0/modules/libmuse_widgets.so: undefined symbol: _ZN7MusEGui11plusSVGIconE

Cheers
kybos

**Comment by terminator356** [comment]
You added icons from our icon library to widgets/scrollscale.cpp 3 days ago:
`scaleUp->setIcon (*plusSVGIcon);`

This would have required linking the widgets library to our icons library.
But this cannot be allowed at the moment, because our icons library is dynamic,
 and depends on themes, requiring a call to initIcons().
If we were ever to share our widgets library outside MusE, a third party app would have to
 not only have access to the complete icon library and the icons, but also it would have to call initIcons().

Sharing the widgets libraries (Awl and widgets) outside MusE is a goal, it was done in the original
 MusE-2 so that users could design MESS plugin UIs and LADSPA UIs in QtDesigner using the
 widgets library which got installed on the system and provided our custom widgets inside QtDesigner.
This would also allow third-party apps to support MESS and our MESS plugins.

For example, Deicsonze uses the Awl library and SimpleDrums uses one of our widgets, the meter.

Currently we don't install the libraries outside MusE, that was disabled, but it was a goal to
 re-activate all that stuff.

Therefore it would be awkward to be forced to install the complete icons library along with 
 our widgets library just because one or more widgets use those icons.
A third-party app would have to call initIcons() and would need access to all the icon files.

So after careful consideration, the best thing for this bug at the moment was to move the
 scrollscale widget into the components folder, where our icons are free to be used.

In addition, the cpu_toolbar widget was also moved.
It uses an icon.
But interestingly, it does so with this line:
`    _resetButton->setIcon(QIcon(":/xpm/cpu.xpm"));`
It bypasses our icons class altogether and loads directly from the resources.
I have moved it into components so that you might change the icon and allow it to conform
 to the icon class theming rules.

**Comment by kybos** [comment]
Thanks Tim.
Still I think we need a real solution to this permanent source of problems, otherwise it will happen again inevitably, as there is apparently no technical check that can catch this kind of issues.
I only incidentally compiled with dynamic linking and only incidentally noticed that simpledrums was missing...

**Comment by spamatica** [comment]
One way (the easy way) is to remove the support for building dynamic libraries. I'm not sure why we started doing it. I recall the link phase used to take a very long time but apart from that I'm not sure what benefits there are? We are still very much a monolithic application.

**Comment by terminator356** [comment]
Actually my (non-static) build stopped on that undefined symbol error.
It should be sufficient to stop anyone from attempting to add icons to the widget library.
I should put a warning right there in the widgets cmake file, that no one should link with the icons library,
 in case someone thinks it wise to do so.

I'm not sure why kybos's build did not fail like mine did, only his SimpleDrums failed to load.
I'm thinking maybe because his SimpleDrums or widgets library file was already built and installed,
 so the compiler didn't bother warning ? Something of that nature ?

Removing the dynamic build feature would not help.
In static mode you would (should) still get an undefined symbol error, requiring you to
 link the icons library into the widgets library, which is forbidden for now.

In fact, the widgets library really ought to be dynamic only, in case we ever share it some day.
You'll notice that I already purposely have made several of our libraries **dynamic only**.
Well, actually, during this fix I looked at installing **both dynamic and static** versions
 of some of our libraries, especially ones that we might someday share externally.
In other words, just like a MusE development package, like many projects do.
A MusE development package would contain the MESS library, the widgets library,
 and a couple more necessities. Ideally both dynamic and static libraries are included.
Yes possibly even include the icons library - if we decide to somehow use our icons in the widgets library,
 it will be required. So I was kind of trying to avoid that, requiring an icons library seems a bit weird,
 especially since the library must be initialized with a call to initIcons(), and theme paths must be supplied.

You'll notice the Awl widgets library has no icons.
Deicsonze synth links with Awl, and there's no problem with widget icons.

So how might we allow icons in our widgets, then, without using our icons library?
The only thing I can think of is that we have separate, normal, simple, non-dynamic icons just for
 those widgets that need them, and we put them right into the widgets library and maybe put them
 in the widgets folder, so that the widgets library is completely 'self-contained', requiring no linking 
to our icons library. Something like that...

Weekend... Man I really need a break... Hopefully get some more programming done.
Cheers.
'The snail'

**Comment by kybos** [comment]
It's good we are prepared for the times where MusE will rule the world (of DAWs at least) and the MESS plugin API will replace LV2 and VST and everybody will want to implement it ;-).
Frankly, I think we are only making life difficult for ourselves without any practical reason with these restrictions.
IMO, SimpleDrums (like any other plugin) should be self-contained and have no dependencies on the core functionality (like widgets).
Nevertheless, this dependency is currently there and it causes problems, so I added the necessary build options to the Compatibility build - the build will now fail if icons are added to the widgets lib or similar things. You won't necessarily notice such things during normal development because it only seems to kick in if you compile from scratch.

**Comment by terminator356** [comment]
The widgets are not the problem. The act of linking widgets with our icons is the problem.
Most plugins need widgets. For example they use Qt or Gtk, or a self contained **static** UI library
 which is very popular these days since statically linking with Qt or Gtk is basically out of the question.

Self contained? Yeah, that's why I made the widgets library stand-alone and self-contained.
The minute we start adding icons from our icons library to the widgets, the widgets are no longer
 self-contained. If our icons library was truly stand-alone and did not require initialization and
 paths to themes, we might be able to use them.
Even then, the widgets would still need a path or resource to the icon files.
This is why the Deicsonze plugin uses the Awl widgets library which uses **no** icons at all.
Deicsonze itself **does** use several icons, but they are part of the plugin and found in its source tree.

If we wish to use our icons in our widgets, one solution is to take the one plugin, SimpleDrums,
 that uses our widgets library, and uses only one freakin' widget from it - the Meter - and construct
 a stand-alone Meter outside of our widgets library. Meter doesn't even use any icons, yet it gets
 dragged into this mess by being part of the widgets library and if the widgets library links with
 our icons library, well, that's where the trouble starts.

Unfortunately, the Meter widget is rather complex, so recreating it outside of our widgets library
 is not really desirable since both codes would need to be synchronized whenever Meter code changes.
I considered doing that during the last fix but decided it was not desirable.

Moral of the story: Don't add our icons to our widgets, until we can straighten this out.
That's why I had to move two of the widgets out of widgets and into components,
 so that the widgets library is self-contained again.
In the components folder, anything goes. You can add icons and so on.
So if you want to add icons to some of our widgets, then move those widgets into the components folder.

**Comment by kybos** [comment]
I think it's not only the icons library. If I remember right, in your case (with the Rasterizer) the problem was not the icons, but a reference to another object from another core library. 
I saw too  that it was the Meter in simpledrums that caused this dependency problem, and I also see that it would be a nuisance to have to double-maintain it. Perhaps replace it by something simpler? I'm sure it uses only a small part of the functionality of the complex Meter control. Probably too much work still...

So I guess we will have to leave it at that for the time being. As mentioned before, now the Compatibility build action should always fail if such a dependency is introduced inadvertently in the future.
